back/app/services/side_fx_user_service.rb

```diff
@@ -32,64 +33,75 @@ def after_create user, current_user
     end
     UpdateMemberCountJob.perform_later
 
-    gained_roles(user).each do |role|
-      case role["type"]
-      when 'project_moderator'
-        SideFxModeratorService.new.after_create(user, Project.find(role["project_id"]), current_user)
-      when 'admin'
-        LogActivityJob.set(wait: 5.seconds).perform_later(user, 'admin_rights_given', current_user, user.updated_at.to_i)
-      end
-    end
-
-    lost_roles(user).each do |role|
-      case role["type"]
-      when 'project_moderator'
-        SideFxModeratorService.new.after_destroy(user, Project.find(role["project_id"]), current_user)
-      when 'admin'
-        user.assigned_initiatives
-          .update_all(assignee_id: nil, updated_at: DateTime.now)
-      end
-    end
+    roles_side_fx(current_user, user)
   end
 
-  def after_destroy frozen_user, current_user
+  def after_destroy(frozen_user, current_user)
     serialized_user = clean_time_attributes(frozen_user.attributes)
-    LogActivityJob.perform_later(encode_frozen_resource(frozen_user), 'deleted', current_user, Time.now.to_i, payload: {user: serialized_user})
+    LogActivityJob.perform_later(encode_frozen_resource(frozen_user), 'deleted', current_user, Time.now.to_i,
+                                 payload: { user: serialized_user })
     UpdateMemberCountJob.perform_later
   end
 
   private
 
-  def lost_roles  user
-    if user.roles_previously_changed?
-      old_roles, new_roles = user.roles_previous_change
-      (old_roles || []) - new_roles
-    else
-      []
+  def roles_side_fx(current_user, user)
+    gained_roles(user).each { |role| role_created_side_fx(role, user, current_user) }
+    lost_roles(user).each { |role| role_destroyed_side_fx(role, user, current_user) }
+  end
+
+  def role_created_side_fx(role, user, current_user)
+    case role['type']
+    when 'project_moderator' then new_project_moderator(role, user, current_user)
+    when 'admin' then new_admin(user, current_user)
     end
   end
 
-  def gained_roles user
-    if user.roles_previously_changed?
-      old_roles, new_roles = user.roles_previous_change
-      new_roles - (old_roles || [])
-    else
-      []
+  def new_project_moderator(role, user, current_user)
+    LogActivityJob.set(wait: 5.seconds).perform_later(
+      user, 'project_moderation_rights_given',
+      current_user, Time.now.to_i,
+      payload: { project_id: role['project_id'] }
+    )
+  end
+
+  def new_admin(user, current_user)
+    LogActivityJob
+      .set(wait: 5.seconds)
+      .perform_later(user, 'admin_rights_given', current_user, user.updated_at.to_i)
+  end
+
+  def role_destroyed_side_fx(role, user, current_user)
+    case role['type']
+    when 'project_moderator' then project_moderator_destroyed(role, user, current_user)
     end
   end
 
-end
+  def project_moderator_destroyed(_role, user, current_user)
+    LogActivityJob.perform_later(user, 'project_moderation_rights_removed', current_user, Time.now.to_i)
+  end
```

Comment:
Most of the code related to project moderator in this file will move to the project management engine in a future PR.